Adding Source Operation example (#1826)
platformsh/platformsh-docs: docs/src/configuration/app/source-operations.md

## Source Operations usage examples

### Update a site from an upstream repository or template

The following source operation syncronizes your branch with an upstream Git repository.

1. Add a project-level variable named `env:UPSTREAM_REMOTE` with the Git URL of the upstream repository. That will make that repository available as a Unix environment variable in all environments, including in the Source Operations environment.

- Variable name: `env:UPSTREAM:REMOTE`
- Variable example value: `https://github.com/platformsh/platformsh-docs`

2. In your  `.platform.app.yaml` file, define a Source Operation to fetch from that upstream repository:

```
source:
    operations:
        upstream-update:
            command: |
                set -e
                git remote add upstream $UPSTREAM_REMOTE
                git fetch --all
                git merge upstream/master
```

> Now every time you run `platform source-operation:run upstream-update` using the CLI on a given branch, the branch fetches all changes from the upstream git repository and then merges the latest changes from the `main` branch in the upstream repository.

Reviewer: ```suggestion
Now every time you run `platform source-operation:run upstream-update` using the CLI on a given branch, the branch fetches all changes from the upstream git repository and then merges the latest changes from the `master` branch in the upstream repository.
```

If there’s a conflict merging from the upstream repository, the source operation will fail and not update from upstream.

Run the `upstream-update` operation on a Development environment rather than directly on Production.

### Revert to the last commit

> The following source operation will revert the last commit pushed to the Git repository. This can be useful if you did not properly test the changes of another operation, and you need to quickly revert to the previous state.

Reviewer: ```suggestion
The following Source Operation will revert the last commit pushed to the Git repository. This can be useful if you did not properly test the changes of another operation, and you need to quickly revert to the previous state.
```

In your  `.platform.app.yaml` file, define a Source Operation to revert the last commit:

```
source:
    operations:
        revert:
            command: |                
                git reset --hard HEAD~
```

Now every time you run `platform source-operation:run revert` using the CLI on a given branch, the operation will revert the last commit pushed to that branch.

### Update Drupal Core

The following source operation will use Composer to update Drupal Core.

1. In your  `.platform.app.yaml` file, define a Source Operation to update Drupal Core:

```
source:
    operations:
        update-drupal-core:
            command: |
                set -e
                composer update drupal/core --with-dependencies
                git add composer.lock
                git commit -m "Automated Drupal Core update."
```

> The Composer command takes the followin parameter:

Reviewer: ```suggestion
The Composer command takes the following parameter:
```

- `--with-dependencies`: use this parameter to also update Drupal Core dependencies. Read more on how to [update Drupal Core via Composer on Drupal.org](https://www.drupal.org/docs/updating-drupal/updating-drupal-core-via-composer).

Now every time you run `platform source-operation:run update-drupal-core` using the CLI on a given branch, the operation will update Drupal Core.

### Download a Drupal extension

> The following source operation will download a Drupal extension.

Reviewer: ```suggestion
The following Source Operation will download a Drupal extension. You can define the Drupal extension by setting an `$EXTENSION` variable or overriding the `$EXTENSION` variable when running the Source Operation.
```

1. In your  `.platform.app.yaml` file, define a Source Operation to update Drupal Core:

```
source:
    operations:
        download-drupal-extension:
            command: |
                set -e
                composer require $EXTENSION
                git add composer.json
                git commit -am "Automated install of: $EXTENSION via Composer."
```

The Composer command takes the following parameter:

- `--with-dependencies`: use this parameter to also update Drupal Core dependencies. Read more on how to [update Drupal Core via Composer on Drupal.org](https://www.drupal.org/docs/updating-drupal/updating-drupal-core-via-composer).

Now every time you run `platform source-operation:run download-drupal-extension --variable env:EXTENSION=drupal/token` using the CLI on a given branch, the operation will download the `drupal/token` on that branch.

Note that the extension is not yet installed, so you need to enable it via the Drupal management interface or using drush.

## External integrations

> If your project is using an external Git integration, the commits resulting from the Source Operations will be pushed to the external repository, before triggering the integration to redeploy the environment.

Reviewer: ```suggestion
If your project is using an external Git integration, any new commits resulting from updating your branch using a Source Operation, will be pushed to that external Git repository first, and then the integration updates Platform.sh and redeploys the environment.
```

> When using an external Git integration, you can not run source operations on environments created from pull or merge requests created on the external repository. 

Reviewer: ```suggestion
When using an external Git integration, you can not run Source Operations on environments created from pull or merge requests created on the external repository. 
```

> If you try running a source operation over a non-supported environment, the following error will be triggered:

Reviewer: ```suggestion
If you try running a Source Operation on a non-supported environment, the following error will be triggered:
```

Also, if multiple applications in a single project both result in a new commit, that will appear as two distinct commits in the Git history but only a single new build/deploy cycle will occur. If multiple applications define source operations with the same name, they will all be executed sequentially on each application.

> ## Examples of source operations

Reviewer: ```suggestion
## Source Operations usage examples
```

### Update a site from an upstream repository or template

> This example will synchronize your branch with the content from an upstream Git repository.

Reviewer: ```suggestion
The following source operation syncronizes your branch with an upstream Git repository.
```

> You need to add a project-level variable named `env:UPSTREAM_REMOTE` with the Git URI of the upstream repository. That will make that repository available as a Unix environment variable in all environments, including in the Source Operations environment.

Reviewer: ```suggestion
1. Add a project-level variable named `env:UPSTREAM_REMOTE` with the Git URL of the upstream repository. That will make that repository available as a Unix environment variable in all environments, including in the Source Operations environment.

- Variable name: `env:UPSTREAM:REMOTE`
- Variable example value: `https://github.com/platformsh/platformsh-docs`
```

Reviewer: For example: 
![image](https://user-images.githubusercontent.com/3729517/121389665-f3358300-c94c-11eb-8325-e43eb2342f75.png)

> Then, define a Source Operation to that upstream repository:

Reviewer: ```suggestion
2. In your  `.platform.app.yaml` file, define a Source Operation to fetch from that upstream repository:
```

@chadwcarlson  ^^ do we have a preferred way to write our YAML files?

```
source:
    operations:
        upstream-update:
            command: |
                set -e
                git remote add upstream $UPSTREAM_REMOTE
                git fetch --all

>                 git merge upstream/master

Reviewer: ```suggestion
                git merge upstream/main
```

Author: Today, all of our templates still all have their default branch called master. And that's the case with most customer upstream repos as well.

Reviewer: I made it main to match the explanation below, the example is okay as long as it's consistent.

```

> Every time the `upstream-update` source operation is ran on a branch, that branch will get checked out, then add a Git remote for the upstream Git repository, then merge the latest changes from its main branch (called `master` in this example). 

Reviewer: ```suggestion

Now every time you run `platform source-operation:run upstream-update` using the CLI on a given branch, the branch fetches all changes from the upstream git repository and then merges the latest changes from the`main` branch in the upstream repository.
```

> Because there are now local changes committed in the local repository, those changes will then get pushed, built, and deployed. If there’s a merge conflict, the script will fail without committing anything.

Reviewer: ```suggestion
If there’s a conflict merging from the upstream repository, the source operation will fail and not update from upstream.
```

I'm not sure what you mean by the local changes - where do those come from @GuGuss ?

Author: What you wrote is correct. Accepted.

> We recommend to run this operation on a Development branch rather than directly on Production.

Reviewer 2: ```suggestion
We recommend running this operation on a Development environment rather than directly on Production.
```

> We recommend running this operation on a Development environment rather than directly on Production.

Reviewer: ```suggestion
Run the `upstream-update` operation on a Development environment rather than directly on Production.
```

## External integrations

> If your project is using an external Git integration, the commits resulting from the Source Operations will be pushed to the external repository, before triggering the integration to redeploy the environment.

Reviewer: I'm also not super clear on what we mean by this @GuGuss

> Only environments created by a branch on the external repository can run source operations. 

Reviewer: ```suggestion

When using an external Git integration, you can not run source operations on environments created from pull or merge requests created on the external repository. 

If you try running a source operation over a non-supported environment, the following error will be triggered:
```

> Note that source oprations can not be run on environments created from pull requests on the external repository (GitHub, Bitbucket, GitLab). Doing so will result in the following error:

Reviewer: ```suggestion
```

I already explained it above :)
